src/anyp/Uri.cc

```diff
@@ -81,6 +84,45 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
```

Comment:
AFAICT, this optimization code attempts to reduce the number of memory (re)allocations in the parsing loop below, but that attempt fails badly on valid input that has no percent characters. Does not such input represent a _dominant_ use case (at least for some of the expected callers)?

Reply:
We don't know, or control it.
The current behaviour mimics the one of the function it aims to replace, but it's also possible that SBuf's heuristics work well enough, at the cost of multiple data copies on encoded-heavy strings.
An optimization that might be interesting is to flip a flag if there is any decoding taking place at all, and if there is none just return the input SBuf. This could help with small optimizations elsewhere (e.g. SBuf::operator =), or by a slightly increased probability of sharing the backing store

Reply:
> We don't know, or control it.

That sounds like an argument for _not_ adding potentially expensive optimizations!

> The current behaviour mimics the one of the function it aims to replace.

Which one? rfc1738_unescape() does not have this behavior.

> Why do you say it fails badly in case of no percent characters?

I say that because the proposed reserveSpace() call leads to an extra buffer allocation and an extra memcpy() in those cases. Without that "optimization", SBuf::append() does not allocate and does not copy the buffer when appending to virgin SBufs. It uses cheap reference counting instead.

I removed that "optimization" in commit 6d70432, addressing this change request.

> if anything we shouldn't call this "worst case"

Yes, that phrasing was not ideal. It is now gone (as a side effect of other changes).

> One possible optimization could be to have a bool switch set to false once per decoding loop, and flipped to true if an encoded triplet is found. Then we can use this information to return the input SBuf instead of the output one. We would still have iterated and copied to output, but thus would allow some optimizations downstream.
> 
> Alternatively, we could special-case and optimize the not-encoded case by adding this code block before the decoding loop, this should essentially make the not-encoded case zero-copy (this has been tested and works):
> 
> ```
>     if (tok.prefix(token, unencodedOctets)) { //[1]
>         if (tok.atEnd())
>             return buf;
>         else
>             output.append(token); //[2]
>     }
> ```
> 
> On top of this, another possible optimization (at the cost of complexity) is to have the output.reserveSpace call happen on just before line marked [2] here, and on the else-branch of the condition at [1] which is needed because if the first character of the encoded string is %, prefix() will return false. This would make the not-encoded case zero-copy and minimal-allocation

I believe the current (as of branch commit 28dc639) implementation is simpler and at least as efficient as the ones discussed above. If your tests reveal unexpected allocations or copies in the no-percent cases, let's revisit this topic.